File: prototype/prototype_pattern.py
import copy

class Prototype:
    def clone(self):
        return copy.copy(self)

class ConcretePrototype(Prototype):
    def __init__(self, name):
        self.name = name

    def set_name(self, name):
        self.name = name

    def __str__(self):
        return f'ConcretePrototype{{name="{self.name}"}}'

class PrototypeManager:
    def __init__(self):
        self.prototypes = {}

    def add_prototype(self, key, prototype):
        self.prototypes[key] = prototype

    def get_prototype(self, key):
        return self.prototypes[key].clone()

if __name__ == "__main__":
    manager = PrototypeManager()

    prototype1 = ConcretePrototype("Prototype 1")
    prototype2 = ConcretePrototype("Prototype 2")
    manager.add_prototype("P1", prototype1)
    manager.add_prototype("P2", prototype2)

    cloned_prototype1 = manager.get_prototype("P1")
    cloned_prototype2 = manager.get_prototype("P2")

    print("Original Prototype 1:", prototype1)
    print("Cloned Prototype 1:", cloned_prototype1)

    print("Original Prototype 2:", prototype2)
    print("Cloned Prototype 2:", cloned_prototype2)

File: prototype/prototype_pattern.js
class Prototype {
    clone() {
        return Object.assign({}, this);
    }
}

class ConcretePrototype extends Prototype {
    constructor(name) {
        super();
        this.name = name;
    }

    setName(name) {
        this.name = name;
    }

    toString() {
        return `ConcretePrototype{name="${this.name}"}`;
    }
}

class PrototypeManager {
    constructor() {
        this.prototypes = {};
    }

    addPrototype(key, prototype) {
        this.prototypes[key] = prototype;
    }

    getPrototype(key) {
        return this.prototypes[key].clone();
    }
}

const manager = new PrototypeManager();

const prototype1 = new ConcretePrototype("Prototype 1");
const prototype2 = new ConcretePrototype("Prototype 2");
manager.addPrototype("P1", prototype1);
manager.addPrototype("P2", prototype2);

const clonedPrototype1 = manager.getPrototype("P1");
const clonedPrototype2 = manager.getPrototype("P2");

console.log("Original Prototype 1:", prototype1.toString());
console.log("Cloned Prototype 1:", clonedPrototype1.toString());

console.log("Original Prototype 2:", prototype2.toString());
console.log("Cloned Prototype 2:", clonedPrototype2.toString());
